src/backend/access/aocs/aocsam.c

```diff
@@ -1199,21 +1211,30 @@ aocs_gettuple(AOCSScanDesc scan, int64 targrow, TupleTableSlot *slot)
 			}
 		}
 
-		if (ds->blockRowCount <= 0)
+		if (scan->segrowsprocessed == 0)
```

Comment:
Why have we changed the condition?

Reply:
If the scan changes to another segment file, we should always read a new block from that file. `scan->segrowsprocesses` is reset to zero during the switch, but `ds->blockRowCount` isn't, so it was possible for it to try to reuse the old block even though we have already switched to a new segment.